great_expectations/core/batch.py

```diff
@@ -574,3 +587,180 @@ def head(self, n_rows=5, fetch_all=False):
+    batch_identifiers: Optional[dict] = None,
+    limit: Optional[int] = None,
+    index: Optional[Union[int, list, tuple, slice, str]] = None,
+    custom_filter_function: Optional[Callable] = None,
+    batch_spec_passthrough: Optional[dict] = None,
+    sampling_method: Optional[str] = None,
+    sampling_kwargs: Optional[dict] = None,
+    splitter_method: Optional[str] = None,
+    splitter_kwargs: Optional[dict] = None,
+    runtime_parameters: Optional[dict] = None,
+    query: Optional[str] = None,
+    path: Optional[str] = None,
+    batch_filter_parameters: Optional[dict] = None,
+    **kwargs,
+) -> Union[BatchRequest, RuntimeBatchRequest]:
+    """Obtain formal BatchRequest typed object from allowed attributes (supplied as arguments).
+    This method applies only to the new (V3) Datasource schema.
+
+    Args:
+        batch_request
+
+        datasource_name
+        data_connector_name
+        data_asset_name
+
+        batch_request
+        batch_data
+        query
+        path
+        runtime_parameters
+        data_connector_query
+        batch_identifiers
+        batch_filter_parameters
+
+        limit
+        index
+        custom_filter_function
+
+        sampling_method
+        sampling_kwargs
+
+        splitter_method
+        splitter_kwargs
+
+        batch_spec_passthrough
+
+        **kwargs
+
+    Returns:
+        (BatchRequest) The formal BatchRequest object
+    """
+
+    if batch_request:
+        if not isinstance(batch_request, BatchRequest):
+            raise TypeError(
+                f"batch_request must be an instance of BatchRequest object, not {type(batch_request)}"
+            )
+        datasource_name = batch_request.datasource_name
+
+    # ensure that the first parameter is datasource_name, which should be a str. This check prevents users
+    # from passing in batch_request as an unnamed parameter.
+    if not isinstance(datasource_name, str):
+        raise ge_exceptions.GreatExpectationsTypeError(
+            f"the first parameter, datasource_name, must be a str, not {type(datasource_name)}"
+        )
+
+    if len([arg for arg in [batch_data, query, path] if arg is not None]) > 1:
+        raise ValueError("Must provide only one of batch_data, query, or path.")
+
+    if any(
+        [
+            batch_data is not None
+            and runtime_parameters
+            and "batch_data" in runtime_parameters,
+            query and runtime_parameters and "query" in runtime_parameters,
+            path and runtime_parameters and "path" in runtime_parameters,
+        ]
+    ):
+        raise ValueError(
+            "If batch_data, query, or path arguments are provided, the same keys cannot appear in the "
+            "runtime_parameters argument."
+        )
+
+    if batch_request:
+        # TODO: Raise a warning if any parameters besides batch_requests are specified
+        return batch_request
+
+    if any([batch_data is not None, query, path, runtime_parameters]):
```

Comment:
Is `is not None` required for the `batch_data` term? If so why is it required for `batch_data` and not the other parameters?

Reply:
Discussed in a call with @NathanFarmer and resolved for now.